### PR TITLE
Implement LeftCensoredDistribution and RightCensoredDistribution

### PR DESCRIPTION
**Pull request by vanAmsterdam**
Implements `LeftCensoredDistribution` and `RightCensoredDistribution`

These distributions take any `base_dist` with a `cdf` method implemented and allow for handling left-or-right censored data. See Issue #1909.

Current status

- [x] implementation
- [x] standard tests
- [ ] censored distribution specific tests
- [ ] updated tutorial / notebook

### CODE REVIEW

File: numpyro/distributions/censored.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    ) -> ArrayLike:
+        return self.base_dist.sample(key, sample_shape)
+
+    @constraints.dependent_property(is_discrete=False, event_dim=0)
+    def support(self) -> ConstraintT:
+        return self._support
+
+    @validate_sample
+    def log_prob(self, value: ArrayLike) -> ArrayLike:
+        minval = 1e-7
```
Comment (vanAmsterdam) [inline numpyro/distributions/censored.py]: this minval was needed to not get -inf in the subsequent `log(1-cdf(x))` calculation for observations with very high tail censored observations; this is what I'm most unsure about

Comment (juanitorduz) [reply numpyro/distributions/censored.py]: Maybe using something like 

```python
    dtype = jnp.result_type(value, float)
    eps = jnp.finfo(dtype).eps
    minval = jnp.sqrt(eps)
```
?

File: numpyro/distributions/censored.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+    def __init__(
+        self,
+        base_dist: DistributionT,
+        censored: ArrayLike = False,
+        *,
+        validate_args: Optional[bool] = None,
+    ):
+        # test if base_dist has an implemented cdf method
+        assert hasattr(base_dist, "cdf")
```
Comment (juanitorduz) [inline numpyro/distributions/censored.py]: Maybe a more informative error:

```python
    if not hasattr(base_dist, "cdf"):
        raise TypeError(
            f"{type(base_dist).__name__} does not have a 'cdf' method. "
            "Censored distributions require a base distribution with an "
            "implemented cumulative distribution function."
        )
    
    # Optionally test that cdf actually works (in validate_args mode)
    if validate_args:
        try:
            test_val = base_dist.support.feasible_like(jnp.array(0.))
            _ = base_dist.cdf(test_val)
        except (NotImplementedError, AttributeError) as e:
            raise TypeError(
                f"{type(base_dist).__name__}.cdf() is not properly implemented."
            ) from e
```
?

File: numpyro/distributions/censored.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        )
+        self.censored = jnp.array(
+            promote_shapes(censored, shape=batch_shape)[0], dtype=jnp.bool
+        )
+        self._support = base_dist.support
+        super().__init__(batch_shape, validate_args=validate_args)
+
+    def sample(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
+    ) -> ArrayLike:
```
Comment (juanitorduz) [inline numpyro/distributions/censored.py:114]: maybe improve type hints ? Distributions have sample signature:

```python
    def sample(
        self, key: Optional[jax.dtypes.prng_key], sample_shape: tuple[int, ...] = ()
    ) -> ArrayLike:
```

Currently, MyPy is not checking these files, if you want them to be checked (and feel like fighting with mypy), you can add the "numpyro.distributions.censored.*", module in https://github.com/pyro-ppl/numpyro/blob/8fec8a1444adbb42c1a409f1fffdfabb63eaae09/pyproject.toml#L120

File: numpyro/distributions/censored.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+from numpyro._typing import ConstraintT, DistributionT
+from numpyro.distributions import constraints
+from numpyro.distributions.distribution import Distribution
+from numpyro.distributions.util import (
+    promote_shapes,
+    validate_sample,
+)
+
+
+class LeftCensoredDistribution(Distribution):
```
Comment (juanitorduz) [inline numpyro/distributions/censored.py:21]: What about using a base class as 

```python
class _BaseCensoredDistribution(Distribution):
    """Base class for censored distributions."""
    
    def __init__(self, base_dist, censored, *, validate_args=None):
        if not hasattr(base_dist, "cdf"):
            raise TypeError(f"{type(base_dist).__name__} must have a cdf method")
        
        batch_shape = lax.broadcast_shapes(base_dist.batch_shape, jnp.shape(censored))
        self.base_dist = jax.tree.map(
            lambda p: promote_shapes(p, shape=batch_shape)[0], base_dist
        )
        self.censored = jnp.array(
            promote_shapes(censored, shape=batch_shape)[0], dtype=jnp.bool
        )
        self._support = base_dist.support
        super().__init__(batch_shape, validate_args=validate_args)
```

File: numpyro/distributions/censored.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+    @constraints.dependent_property(is_discrete=False, event_dim=0)
+    def support(self) -> ConstraintT:
+        return self._support
+
+    @validate_sample
+    def log_prob(self, value: ArrayLike) -> ArrayLike:
+        minval = 1e-7
+
+        def logS(x):
```
Comment (juanitorduz) [inline numpyro/distributions/censored.py]: Suggestion:

The internal functions `logF` and `logS` are concise but could be more descriptive:
- `logF` → `log_cdf_censored`
- `logS` → `log_survival_censored`

File: test/test_distributions.py [hunk 6]
```diff
@@ -32,6 +32,10 @@
     transforms,
 )
 from numpyro.distributions.batch_util import vmap_over
+from numpyro.distributions.censored import (
+    LeftCensoredDistribution,
```
Comment (juanitorduz) [inline test/test_distributions.py:36]: Maybe add some additional custom tests

a) **Gradient tests**: Ensure gradients flow correctly through censored observations
```python
def test_censored_gradient():
    def loss(loc):
        base_dist = dist.Normal(loc, 1.0)
        censored_dist = dist.LeftCensoredDistribution(base_dist, censored=1)
        return -censored_dist.log_prob(2.0)
    
    grad_fn = jax.grad(loss)
    grad = grad_fn(0.0)
    assert jnp.isfinite(grad)
```

b) **Batch dimension broadcasting**: More comprehensive tests with mixed batch shapes
```python
def test_censored_broadcasting():
    # loc: (3,), censored: (2, 1) -> batch_shape: (2, 3)
    base_dist = dist.Normal(jnp.array([0., 1., 2.]), 1.0)
    censored = jnp.array([[0], [1]])
    dist_obj = dist.LeftCensoredDistribution(base_dist, censored)
    assert dist_obj.batch_shape == (2, 3)
```

c) **Sampling correctness**: Verify that samples from censored distributions are valid
```python
def test_censored_sample_validity():
    base_dist = dist.Normal(0., 1.)
    censored_dist = dist.RightCensoredDistribution(base_dist, censored=0)
    samples = censored_dist.sample(random.PRNGKey(0), (1000,))
    # Samples should match base distribution when censored=0
    assert samples.shape == (1000,)
```

File: numpyro/distributions/censored.py (hunk 1 above)
Comment (vanAmsterdam) [reply numpyro/distributions/censored.py]: thanks for this suggestion; I had something similar in before but it failed for in32 dtypes (for discrete distributions like poisson), but jnp.result_type fixed this!

File: numpyro/distributions/censored.py (hunk 2 above)
Comment (vanAmsterdam) [reply numpyro/distributions/censored.py]: I copied this suggestion in the new code

File: numpyro/distributions/censored.py (hunk 3 above)
Comment (vanAmsterdam) [reply numpyro/distributions/censored.py:114]: I'm happy to use your proposed type hint, but I don't see it used consistently in the other distributions? e.g. continuous.py or discrete.py?

File: numpyro/distributions/censored.py (hunk 4 above)
Comment (vanAmsterdam) [reply numpyro/distributions/censored.py:21]: the init signature for IntervalCensored is different (has a left_censored and right_censored) argument; maybe defer this base class to when it's clear how to implement intervalcensored?

another possibility would be to create a 'flexible' CensoredDistribution class where the user can specify left_censored and/or right_censored and the result will be whatever is specified. In fact, the IntervalCensoredDistribution implementation does this already as observations can be left, right, interval or double censored

File: numpyro/distributions/censored.py (hunk 3 above)
Comment (juanitorduz) [reply numpyro/distributions/censored.py:114]: I know, we are working on it :/ https://github.com/pyro-ppl/numpyro/issues (that is for now I suggest this as optional)

File: numpyro/distributions/censored.py (hunk 4 above)
Comment (juanitorduz) [reply numpyro/distributions/censored.py:21]: So maybe for now we leave it as it is and wait for the `IntervalCensoredDistribution` in a following PR?

File: test/test_distributions.py (hunk 6 above)
Comment (vanAmsterdam) [reply test/test_distributions.py:36]: thanks for these suggestions on additional tests; I've adapted them and added them

File: numpyro/distributions/censored.py (hunk 4 above)
Comment (vanAmsterdam) [reply numpyro/distributions/censored.py:21]: I'd say yes

**Comment by vanAmsterdam** [comment]
> You can try ruining `make doctest` locally to make the debugging easier :)

yes figured this one out a little too late :) but it's passing now so I think we're set for a review. thanks for the help!

**Review by fehiepsi — CHANGES_REQUESTED** [review]
Actually, I have some minor  comments.

File: numpyro/distributions/censored.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        self.censored = jnp.array(
+            promote_shapes(censored, shape=batch_shape)[0], dtype=jnp.bool
+        )
+        self._support = base_dist.support
+        super().__init__(batch_shape, validate_args=validate_args)
+
+    def sample(
+        self, key: Optional[jax.dtypes.prng_key], sample_shape: tuple[int, ...] = ()
+    ) -> ArrayLike:
+        return self.base_dist.sample(key, sample_shape)
```
Comment (fehiepsi) [inline numpyro/distributions/censored.py]: I think you need `self.base_dict.expand(self.batch_shape).sample(...)`.

File: numpyro/distributions/censored.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
```
Comment (fehiepsi) [inline numpyro/distributions/censored.py]: You'll need self.base_dict.expand(self.batch_shape).sample(...). promote_shapes does not broadcast the parameters, it only adds some missing singleton dimensions.

Comment (vanAmsterdam) [reply numpyro/distributions/censored.py]: good catch; updated in newest commit and also added a test

**Review by fehiepsi — APPROVED** [review]
